rholang/src/main/rholang/cross-shard.rho

```diff
@@ -0,0 +1,538 @@
+                } |
+                for (@true <- successCh) {
+                  if (balance < depositFee) {
+                    statusOut!(["failure", "Not enough funds to pay the fee."])
+                  } else {
+                    // Tell child to mint a wallet containing (balance - depositFee) guarded by @targetPubKey
+                    @[mailBoxClient, "send"]!([balance - depositFee, targetPubKey], *statusOut)
+                  }
+                }
+              }
+            }
+          }
+        } |
+        
+        contract @[*Depository, "withdraw"](@pubKey, @uuid, @amount, @targetPubKey, statusOut, @sig) = {
+          @[mailBoxClient, "receive"]!(pubKey, uuid, [amount, targetPubKey], *statusOut, sig)
+        } |
+
+        contract @[*Depository, "updateParams"](@pubKey, @pubKeySet, @sendQuorumSize, @cancelQuorumSize, statusOut, @sig) = {
+          @[mailBoxAdmin, "updateParams"]!(pubKey, pubKeySet, sendQuorumSize, cancelQuorumSize, *statusOut, sig)
+        } |
+        
+        contract withdrawInternal(@[amount, algorithm, targetPubKey]) = {
+          // create wallet from depository for amount at a new uuid guarded by the public key
+          new splitPurseOut in {
+            @[depositoryPurse, "split"]!(amount, *splitPurseOut) |
+            // TODO(pyrocto): failure case?
+            for (@[purse] <- splitPurseOut) {
+              new walletOut in {
+                // TODO(pyrocto): make algorithm parametric
+                @"BasicWallet"!(purse, "ed25519", targetPubKey, *walletOut) |
+                for (@wallet <- walletOut) {
+                  @[*r, "register"]!(`rho:uuid`, bundle+{wallet}, Nil)
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  } |
+  
+  contract @[*MintFactory, "create"](@pubKeySet, @sendQuorumSize, @cancelQuorumSize, @depositFee, mintOut) = {
+    new Mint, mintInternal, mailBoxOut, underlyingMintOut in {
+      mintOut!(*Mint) |
+      @"MakeMint"!(*underlyingMintOut) |
+      @[*MailBoxFactory, "create"]!(pubKeySet, sendQuorumSize, cancelQuorumSize, *mintInternal, *mailBoxOut) |
+      
+      for (@[mailBoxClient, mailBoxAdmin] <- mailBoxOut) {
+
+        contract @[*Mint, "sendToParent"](@purse, @targetPubKey, statusOut) = {
+          new return in {
+            // Find out how much this purse has.
+            @[purse, "getBalance"]!(*return) |
+            for (@balance <- return) {
+              new splitPurseOut in {
+                // Create new wallet, transfer funds
+                @[purse, "split"]!(balance, *splitPurseOut) |
+                for (@[] <- splitPurseOut) {
+                  statusOut!(["failure", "Could not deposit funds."])
+                } |
+                // Throw away the resulting purse.
+                for (@[_] <- splitPurseOut) {
+                  if (balance < depositFee) {
+                    statusOut!(["failure", "Not enough funds to pay the fee."])
```

Comment:
I don't really like that the money gets destroyed no matter what. It would be more user friendly to check if the balance was sufficient first.